File: app/assets/javascripts/pos/models/sale.js
SwapzPOS.Sale = SwapzPOS.Base.extend({
  complete: null,
  completedAt: null,
  configurable: null,
  certificate_id: null,
  certificate: null,
  customer_id: null,
  customer: null,
  flagged: null,
  lines: null,
  note: null,
  payment: null,
  sku: null,
  taxRate: null,
  till_id: null,
  till: null,
  store_id: null,
  store: null,
  user_id: null,
  user: null,
  init: function() {
    this._super();
    this.set('complete', false);
    this.set('lines', Ember.A());
    this.set('payment', SwapzPOS.Payment.create());
  },
  quantity: function() {
    var lines = this.get('lines');
    var quantity = 0;
    lines.forEach(function(line) {
      if (!line.get('_remove')) {
        quantity += parseInt(line.get('quantity'));
      }
    });
    return quantity;
  }.property('lines', 'lines.@each', 'lines.@each.quantity', 'lines.@each._remove'),
  subtotal: function() {
    var lines = this.get('lines');
    var subtotal = 0;
    lines.forEach(function(line) {
      if (!line.get('_remove')) {
        subtotal += line.get('subtotal');
      }
    });
    return subtotal;
  }.property('taxRate', 'lines', 'lines.@each', 'lines.@each.subtotal', 'lines.@each._remove'),
  subtotalAfterStoreCredit: function() {
    return this.get('subtotal') - this.get('payment.storeCredit');
  }.property('taxRate', 'lines', 'lines.@each', 'lines.@each.subtotal', 'lines.@each._remove', 'payment.storeCredit'),
  taxableSubtotal: function() {
    var lines = this.get('lines');
    var subtotal = 0;
    lines.forEach(function(line) {
      if (line.get('taxable') &&  !line.get('_remove')) {
        subtotal += line.get('subtotal');
      }
    });
    return subtotal;
  }.property('taxRate', 'lines', 'lines.@each', 'lines.@each.subtotal', 'lines.@each.taxable', 'lines.@each._remove'),
  tax: function() {
    var subtotal = this.get('subtotal') - this.get('payment.storeCredit');
    if (subtotal > 0) {
      var taxableSubtotal = this.get('taxableSubtotal') - this.get('payment.storeCredit');
      if (taxableSubtotal > 0) {
        return parseInt(Math.round(taxableSubtotal * this.get('taxRate')));
      } else {
        return 0;
      }
    } else {
      return 0;
    }
  }.property('taxRate', 'lines', 'lines.@each', 'lines.@each.subtotal', 'lines.@each.taxable', 'lines.@each._remove', 'payment.storeCredit'),
  total: function() {
    return this.get('subtotalAfterStoreCredit') + this.get('tax');
  }.property('taxRate', 'lines', 'lines.@each', 'lines.@each.subtotal', 'lines.@each._remove', 'payment.storeCredit'),
  due: function() {
    return this.get('total') - this.get('payment.total');
  }.property('taxRate', 'total', 'payment.storeCredit', 'payment.giftCard', 'payment.check', 'payment.credit', 'payment.cash'),
  dueLabel: function() {
    var due = this.get('due')
    if (due > 0) {
      return "Amount Due:";
    } else {
      return "Change Due:";
    }
  }.property('due'),
  saveable: function() {
    return !this.get('complete');
  }.property('complete'),
  nonSaveable: function() {
    return !this.get('saveable');
  }.property('saveable'),
  completable: function() {
    return !this.get('complete') && this.get('user') && this.get('store') && this.get('till') && this.get('payment') && this.get('quantity') > 0 && this.get('due') <= 0;
  }.property('complete', 'user', 'store', 'till', 'payment', 'quantity', 'due'),
  nonCompletable: function() {
    return !this.get('completable');
  }.property('completable')
});

SwapzPOS.Sale.reopenClass({
  singular: 'sale',
  plural: 'sales'
});

SwapzPOS.Sale.reopen({
  entity: function() {
    var entity = this._super();
    $.extend(entity.sale, {
      complete: this.get('complete'),
      flagged: this.get('flagged'),
      tax_rate: this.get('taxRate'),
      note: this.get('note'),
      certificate_id: this.get('certificate_id'),
      customer_id: this.get('customer_id'),
      till_id: this.get('till_id'),
      user_id: this.get('user_id'),
      store_id: this.get('store_id'),
      lines_attributes: [],
      payment_attributes: null
    });
    if (this.get('sku')) {
      entity.sale.sku = this.get('sku');
    }
    if (this.get('certificate')) {
      entity.sale.certificate_id = this.get('certificate.id');
    }
    if (this.get('customer')) {
      entity.sale.customer_id = this.get('customer.id');
    }
    if (this.get('till')) {
      entity.sale.till_id = this.get('till.id');
    }
    if (this.get('store')) {
      entity.sale.store_id = this.get('store.id');
    }
    if (this.get('user')) {
      entity.sale.user_id = this.get('user.id');
    }
    if (this.get('payment')) {
      entity.sale.payment_attributes = {
        id: this.get('payment.id'),
        cash: this.get('payment.cash'),
        credit: this.get('payment.credit'),
        check: this.get('payment.check'),
        gift_card: this.get('payment.giftCard'),
        store_credit: this.get('payment.storeCredit')
      }
      if (this.get('payment.id')) {
        entity.sale.payment_attributes.id = this.get('payment.id');
      }
    }
    this.get('lines').forEach(function(line) {
      var _line = {
        amount: line.get('amount'),
        amount_cash: line.get('amountCash'),
        amount_credit: line.get('amountCredit'),
        bullets: line.get('bullets'),
        inventory: line.get('inventory'),
        certificate: line.get('certificate'),
        quantity: line.get('quantity'),
        note: line.get('note'),
        sku: line.get('sku'),
        taxable: line.get('taxable'),
        title: line.get('title'),
        _destroy: line.get('_remove')
      };
      if (line.get('id')) {
        _line.id = line.get('id');
      }
      if (line.get('certificate')) {
        _line.certificate_id = line.get('certificate.id');
      }
      if (line.get('unit')) {
        _line.unit_id = line.get('unit.id');
      }
      if (line.get('item')) {
        _line.item_id = line.get('item.id');
      }
      entity.sale.lines_attributes.addObject(_line);
    });
    return entity;
  },
  assign: function(data) {
    this._super(data);
    this.setProperties({
      sku: data.sku,
      complete: data.complete,
      flagged: data.flagged,
      taxRate: data.tax_rate,
      note: data.note,
      certificate_id: data.certificate_id,
      customer_id: data.customer_id,
      till_id: data.till_id,
      user_id: data.user_id,
      store_id: data.store_id,
    });
    if (data.completed_at) {
      this.set('completedAt', new Date(data.completed_at));
    }
    if (data.certificate) {
      this.set('certificate', SwapzPOS.Certificate.create().assign(data.certificate));
    }
    if (data.customer) {
      this.set('customer', SwapzPOS.Customer.create().assign(data.customer));
    }
    if (data.user) {
      this.set('user', SwapzPOS.User.create().assign(data.user));
    }
    if (data.till) {
      this.set('till', SwapzPOS.Till.create().assign(data.till));
    }
    if (data.store) {
      this.set('store', SwapzPOS.Store.create().assign(data.store));
    }
    if (data.payment) {
      this.set('payment', SwapzPOS.Payment.create({
        id: data.payment.id,
        cash: data.payment.cash,
        credit: data.payment.credit,
        check: data.payment.check,
        giftCard: data.payment.gift_card,
        storeCredit: data.payment.store_credit
      }));
    }
    if (data.lines) {
      var lines = [];
      data.lines.forEach(function(_line){
        var line = SwapzPOS.Line.create({
          id: _line.id,
          amount: _line.amount,
          amountCash: _line.amount_cash,
          amountCredit: _line.amount_credit,
          inventory: _line.inventory,
          certificate: _line.certificate,
          quantity: _line.quantity,
          note: _line.note,
          sku: _line.sku,
          taxable: _line.taxable,
          title: _line.title
        });
        if (_line.certificate) {
          line.set('certificate', SwapzPOS.Certificate.create().assign(_line.certificate));
        }
        if (_line.unit) {
          line.set('unit', SwapzPOS.Unit.create().assign(_line.unit));
        }
        if (_line.item) {
          line.set('item', SwapzPOS.Item.create().assign(_line.item));
        }
        line.set('bullets', _line.bullets);
        lines.addObject(line);
      });
      this.set('lines', lines);
    }
    return this;
  }
});

File: app/models/sale.rb
class Sale
  include Mongoid::Document
  include Mongoid::Timestamps
  include Mongoid::Search
  
  field :complete,        :type => Boolean,   :default => false
  field :completed_at,    :type => Time
  field :flagged,         :type => Boolean,   :default => false
  field :note,            :type => String
  field :sku,             :type => String,    :default => ->{ SecureRandom.hex(8).upcase }
  field :tax_rate,        :type => Float,     :default => 0
  field :credit_balance,  :type => Integer,   :default => 0
  
  index({ :account_id => 1, :sku => 1 })
  
  attr_reader :quantity, :subtotal, :subtotal_taxable, :subtotal_after_store_credit, :subtotal_taxable_after_store_credit, :tax, :total, :due
  
  validates_inclusion_of    :complete,  :in => [true, false]
  validates_inclusion_of    :flagged,   :in => [true, false]
  validates_presence_of     :sku
  validates_uniqueness_of   :sku, scope: :account_id
  validates_presence_of     :tax_rate
  validates_inclusion_of    :complete, :in => [true, false]
  validates_presence_of     :payment, :store, :till, :user
  
  default_scope ->{ where(:account_id => Account.current_id) }
  
  scope :search, ->(query) { query ? full_text_search(query) : all }
  scope :sorted, ->(sort) {
    if sort
      sort[:asc] ||= []
      sort[:desc] ||= []
      asc(sort[:asc]).desc(sort[:desc])
    else
      desc(:updated_at)
    end
  }
  scope :user_id, ->(user_id) { where(:user_id => user_id) if user_id }
  scope :customer_id, ->(customer_id) { where(:customer_id => customer_id) if customer_id }
  scope :store_id, ->(store_id) { where(:store_id => store_id) if store_id }
  scope :till_id, ->(till_id) { where(:till_id => till_id) if till_id }
  
  belongs_to  :account
  belongs_to  :certificate
  belongs_to  :customer
  embeds_many :lines,   cascade_callbacks: true
  embeds_one  :payment, cascade_callbacks: true
  belongs_to  :store
  belongs_to  :till
  belongs_to  :user
  
  after_save :_complete
  
  accepts_nested_attributes_for :lines, :allow_destroy => true
  accepts_nested_attributes_for :payment
  
  search_in :sku, :note, :lines => [:title, :sku, :note], :certificate => [:sku], :customer => [:first_name, :last_name, :sku], :till => [:name], :store => [:name], :user => [:username, :email, :first_name, :last_name]
  
  liquid_methods :account, :created_at, :certificate, :credit_balance, :complete, :completed_at, :customer, :flagged, :lines, :payment, :note, :sku, :quantity, :subtotal, :subtotal_taxable, :subtotal_after_store_credit, :subtotal_taxable_after_store_credit, :tax, :tax_rate, :total, :store, :till, :user, :due, :updated_at
  
  def self.trend(now, period_in_days)
    
    map = %Q{
      function() {
        var oneDay = 24*60*60*1000;
        var now = new Date();
        var date = new Date(this.completed_at);

        var days_ago = Math.round(Math.abs((now.getTime() - date.getTime())/(oneDay))) * -1;
        
        var quantity = 0;
        var subtotal = 0;
        var subtotal_taxable = 0;
        var payment = this.payment
        if (this.lines) {
          this.lines.forEach(function(line) {
            quantity += line.quantity;
            subtotal += line.quantity * line.amount;
            if (line.taxable) {
              subtotal_taxable += line.quantity * line.amount;
            }
          });
        }
        var subtotal_after_store_credit = subtotal - payment.store_credit;
        var subtotal_taxable_after_store_credit = subtotal_taxable - payment.store_credit;
        var tax = 0;
        if (subtotal_taxable_after_store_credit > 0) {
          tax += Math.round(subtotal_taxable_after_store_credit * this.tax_rate);
        }
        var total = subtotal_after_store_credit + tax;
        
        emit(days_ago, {count: 1, quantity: quantity, total: total});
      }
    }

    reduce = %Q{
      function(key, values) {
        var result = {count: 0, quantity: 0, total: 0};
        values.forEach(function(value) {
          result.count += value.count;
          result.quantity += value.quantity;
          result.total += value.total;
        });
        return result;
      }
    }
    
    @sales = Sale.where(:complete => true, :completed_at.gte => now.beginning_of_day - period_in_days.to_i.days)
    
    @result = []
    unless @sales.empty?
      @data = @sales.map_reduce(map, reduce).out(inline: true)
      ((period_in_days.to_i * -1)..0).each do |day|
        @point = @data.find{|d| d['_id'] == day}
        if @point
          @result << { day: day, count: @point["value"]["count"], quantity: @point["value"]["quantity"], total: @point["value"]["total"] }
        else
          @result << { day: day, count: 0, quantity: 0, total: 0 }
        end
      end
    else
      ((period_in_days.to_i * -1)..0).each do |day|
        @result << { day: day, count: 0, quantity: 0, total: 0 }
      end
    end
    
    @result
  end
  
  def quantity
    quantity = 0
    lines.each do |line|
      quantity += line.quantity
    end
    quantity
  end
  
  def subtotal
    subtotal = 0
    lines.each do |line|
      subtotal += line.subtotal
    end
    subtotal
  end
  
  def subtotal_taxable
    subtotal_taxable = 0
    lines.each do |line|
      if line.taxable
        subtotal_taxable += line.subtotal
      end
    end
    subtotal_taxable
  end
  
  def subtotal_after_store_credit
    subtotal_after = self.subtotal
    if self.payment
      subtotal_after -= self.payment.store_credit
    end
    subtotal_after
  end
  
  def subtotal_taxable_after_store_credit
    subtotal_taxable_after = self.subtotal_taxable
    if self.payment
      subtotal_taxable_after -= self.payment.store_credit
    end
    subtotal_taxable_after
  end
  
  def tax
    tax = 0
    subtotal_taxable_after_store_credit = self.subtotal_taxable_after_store_credit
    if subtotal_taxable_after_store_credit > 0
      tax += (subtotal_taxable_after_store_credit * self.tax_rate).round
    end
    tax
  end
  
  def total
    self.subtotal_after_store_credit + self.tax
  end
  
  def due
    if self.payment
      self.total - self.payment.total
    else
      self.total
    end
  end
  
  protected
  
  def _complete
    if complete && completed_at.nil?
      if till && payment && user
        title = "Sale #{sku} #{Time.now.to_s} (#{user.fullname})"
        till.adjustments.create(:title => title, :amount => payment.cash + due, :balance => till.balance, :user => user)
      end
      if certificate && payment
        certificate.subtract_balance(payment.gift_card) if payment.gift_card > 0
      end
      if customer && payment
        customer.subtract_credit(payment.store_credit) if payment.store_credit > 0
        set(:credit_balance, customer.credit)
      end
      lines.each do |line|
        if line.unit
          line.unit.subtract_quantity(line.quantity) if line.quantity > 0
        end
        if line.certificate
          line.certificate.update_attributes(:active => true, :customer => customer)
        end
      end
      set(:completed_at, Time.now.utc)
    end
    true
  end
end